File: ios/HybridTextRecognizerResult.swift
//
//  HybridTextRecognizerResult.swift
//  VisionCameraTextScanner
//
//  Created by Alex Korn on 21.04.26.
//

import Vision
import NitroModules

final class HybridTextRecognizerResult: HybridTextRecognizerResultSpec {
  private let observations: [VNRecognizedTextObservation]
  private let _frameWidth: Double
  private let _frameHeight: Double

  init(observations: [VNRecognizedTextObservation], frameWidth: Int, frameHeight: Int) {
    self.observations = observations
    self._frameWidth = Double(frameWidth)
    self._frameHeight = Double(frameHeight)
    super.init()
  }

  var frameWidth: Double { _frameWidth }
  var frameHeight: Double { _frameHeight }

  var text: String {
    return observations.compactMap { observation in
      observation.topCandidates(1).first?.string
    }.joined(separator: " ")
  }

  // Vision Framework uses bottom-left origin with normalized (0–1) coordinates.
  // Convert to top-left origin (UIKit/RN convention) so callers get a consistent system.
  private func visionBoxToBoundingBox(_ box: CGRect) -> BoundingBox {
    return BoundingBox(
      x: box.origin.x,
      y: 1.0 - (box.origin.y + box.size.height),
      width: box.size.width,
      height: box.size.height
    )
  }

  var boundingBox: BoundingBox {
    guard !observations.isEmpty else {
      return BoundingBox(x: 0, y: 0, width: 0, height: 0)
    }
    // Compute union of all observation bounding boxes (Vision coords: bottom-left origin).
    let minX = observations.map { $0.boundingBox.minX }.min()!
    let minY = observations.map { $0.boundingBox.minY }.min()!
    let maxX = observations.map { $0.boundingBox.maxX }.max()!
    let maxY = observations.map { $0.boundingBox.maxY }.max()!
    let unionBox = CGRect(x: minX, y: minY, width: maxX - minX, height: maxY - minY)
    return visionBoxToBoundingBox(unionBox)
  }

  var cornerPoints: [Point] {
    guard !observations.isEmpty else {
      return []
    }
    let b = boundingBox
    return [
      Point(x: b.x,           y: b.y),
      Point(x: b.x + b.width, y: b.y),
      Point(x: b.x + b.width, y: b.y + b.height),
      Point(x: b.x,           y: b.y + b.height),
    ]
  }

  // Vision Framework does not provide a block/line/word hierarchy — each VNRecognizedTextObservation
  // is a single recognized text region. We map each observation to one Block containing one Line
  // containing one Word. This is an intentional simplification; the structure mirrors Android ML Kit's
  // output shape for API consistency.
  var blocks: [TextBlock] {
    return observations.map { observation in
      guard let candidate = observation.topCandidates(1).first else {
        return TextBlock(
          text: "",
          boundingBox: BoundingBox(x: 0, y: 0, width: 0, height: 0),
          cornerPoints: [],
          lines: []
        )
      }

      let b = visionBoxToBoundingBox(observation.boundingBox)
      let corners = [
        Point(x: b.x,           y: b.y),
        Point(x: b.x + b.width, y: b.y),
        Point(x: b.x + b.width, y: b.y + b.height),
        Point(x: b.x,           y: b.y + b.height),
      ]
      return TextBlock(
        text: candidate.string,
        boundingBox: b,
        cornerPoints: corners,
        lines: [
          TextLine(
            text: candidate.string,
            boundingBox: b,
            cornerPoints: [],
            words: [
              TextWord(
                text: candidate.string,
                boundingBox: b,
                cornerPoints: []
              )
            ]
          )
        ]
      )
    }
  }
}

File: android/src/main/java/com/margelo/nitro/camera/textrecognizer/HybridTextRecognizerResult.kt
package com.margelo.nitro.camera.textrecognizer

import com.google.mlkit.vision.text.Text

// ML Kit returns absolute pixel coordinates with top-left origin.
// Normalize to (0–1) range so the JS side gets the same convention as iOS.
class HybridTextRecognizerResult(
  private val result: Text,
  frameWidth: Int,
  frameHeight: Int,
) : HybridTextRecognizerResultSpec() {

  private val _frameWidth: Int = frameWidth
  private val _frameHeight: Int = frameHeight

  private fun android.graphics.Rect.toNormalizedBoundingBox(): BoundingBox = BoundingBox(
    x = left.toDouble() / _frameWidth,
    y = top.toDouble() / _frameHeight,
    width = (right - left).toDouble() / _frameWidth,
    height = (bottom - top).toDouble() / _frameHeight,
  )

  private fun Array<android.graphics.Point>.toNormalizedPoints(): Array<Point> =
    map { Point(it.x.toDouble() / _frameWidth, it.y.toDouble() / _frameHeight) }.toTypedArray()

  override val frameWidth: Double = _frameWidth.toDouble()
  override val frameHeight: Double = _frameHeight.toDouble()

  override val text: String
    get() = result.text

  override val boundingBox: BoundingBox
    get() {
      val rects = result.textBlocks.mapNotNull { it.boundingBox }
      if (rects.isEmpty()) return BoundingBox(0.0, 0.0, 0.0, 0.0)
      val left = rects.minOf { it.left }
      val top = rects.minOf { it.top }
      val right = rects.maxOf { it.right }
      val bottom = rects.maxOf { it.bottom }
      return BoundingBox(
        x = left.toDouble() / _frameWidth,
        y = top.toDouble() / _frameHeight,
        width = (right - left).toDouble() / _frameWidth,
        height = (bottom - top).toDouble() / _frameHeight,
      )
    }

  override val cornerPoints: Array<Point>
    get() {
      val b = boundingBox
      if (b.width == 0.0 && b.height == 0.0) return emptyArray()
      return arrayOf(
        Point(b.x, b.y),
        Point(b.x + b.width, b.y),
        Point(b.x + b.width, b.y + b.height),
        Point(b.x, b.y + b.height),
      )
    }

  override val blocks: Array<TextBlock>
    get() = result.textBlocks.map { block ->
      TextBlock(
        text = block.text,
        boundingBox = block.boundingBox?.toNormalizedBoundingBox() ?: BoundingBox(0.0, 0.0, 0.0, 0.0),
        cornerPoints = block.cornerPoints?.toNormalizedPoints() ?: emptyArray(),
        lines = block.lines.map { line ->
          TextLine(
            text = line.text,
            boundingBox = line.boundingBox?.toNormalizedBoundingBox() ?: BoundingBox(0.0, 0.0, 0.0, 0.0),
            cornerPoints = line.cornerPoints?.toNormalizedPoints() ?: emptyArray(),
            words = line.elements.map { element ->
              TextWord(
                text = element.text,
                boundingBox = element.boundingBox?.toNormalizedBoundingBox() ?: BoundingBox(0.0, 0.0, 0.0, 0.0),
                cornerPoints = element.cornerPoints?.toNormalizedPoints() ?: emptyArray(),
              )
            }.toTypedArray()
          )
        }.toTypedArray()
      )
    }.toTypedArray()
}
